runtime/near-evm-runner/src/utils.rs

```diff
@@ -0,0 +1,317 @@
+            buffer[(1 + 20 + 32)..].copy_from_slice(&code_hash[..]);
+            (From::from(keccak(&buffer[..])), Some(code_hash))
+        }
+        CreateContractAddress::FromSenderAndCodeHash => {
+            let code_hash = keccak(code);
+            let mut buffer = [0u8; 20 + 32];
+            buffer[..20].copy_from_slice(&sender[..]);
+            buffer[20..].copy_from_slice(&code_hash[..]);
+            (From::from(keccak(&buffer[..])), Some(code_hash))
+        }
+    }
+}
+
+#[derive(Eq, PartialEq, Debug, Ord, PartialOrd)]
+pub struct Balance(pub u128);
+
+impl Balance {
+    pub fn from_be_bytes(bytes: [u8; 16]) -> Self {
+        Balance(u128::from_be_bytes(bytes))
+    }
+
+    pub fn to_be_bytes(&self) -> [u8; 16] {
+        self.0.to_be_bytes()
+    }
+}
+
+impl Serialize for Balance {
+    fn serialize<S>(
+        &self,
+        serializer: S,
+    ) -> std::result::Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&format!("{}", &self.0))
+    }
+}
+
+impl<'de> Deserialize<'de> for Balance {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, <D as Deserializer<'de>>::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        u128::from_str_radix(&s, 10).map(Balance).map_err(serde::de::Error::custom)
+    }
+}
+
+impl From<Balance> for u128 {
+    fn from(balance: Balance) -> Self {
+        balance.0
+    }
+}
+
+pub fn format_log(topics: Vec<H256>, data: &[u8]) -> std::result::Result<Vec<u8>, std::io::Error> {
+    let mut result = Vec::with_capacity(1 + topics.len() * 32 + data.len());
+    result.write_u8(topics.len() as u8)?;
+    for topic in topics.iter() {
+        result.write(&topic.0)?;
+    }
+    result.write(data)?;
+    Ok(result)
+}
+
+pub fn near_erc721_domain(chain_id: U256) -> RawU256 {
+    let mut bytes = Vec::with_capacity(70);
+    bytes.extend_from_slice(
+        &keccak("EIP712Domain(string name,string version,uint256 chainId)".as_bytes()).as_bytes(),
+    );
+    bytes.extend_from_slice(b"NEAR");
+    bytes.extend_from_slice(&[0x01]);
+    bytes.extend_from_slice(&u256_to_arr(&chain_id));
+    keccak(&bytes).into()
+}
+
+pub fn method_name_to_rlp(method_name: String) -> [u8; 4] {
+    let mut result = [0u8; 4];
+    result.copy_from_slice(&keccak(method_name)[..4]);
+    result
+}
+
+pub fn prepare_meta_call_args(
+    domain_separator: &RawU256,
+    account_id: &AccountId,
+    nonce: U256,
+    fee_amount: U256,
+    fee_address: Address,
+    contract_address: Address,
+    method_name: &str,
+    args: &[u8],
+) -> RawU256 {
+    let mut bytes = Vec::with_capacity(32 + 32 + 20 + account_id.len() + 4 + args.len());
+    bytes.extend_from_slice(
+        &keccak(
+            "NearTx(string evmId, uint256 nonce, uint256 feeAmount, uint256 feeAddress, address contractAddress, string contractMethod, Arguments arguments)"
```

Comment:
I think when we do this, we don't want the spaces after the commas, yes?

Reply:
you are right, this needs more work in general. Apparently the sub types are also encoded in this string.
I think we need to send a encoded string for this stuff overall, not just method name